core/src/Cabana_ExecutionPolicy.hpp

```diff
@@ -13,53 +13,219 @@
 #define CABANA_EXECUTIONPOLICY_HPP
 
 #include <Cabana_Macros.hpp>
+#include <impl/Cabana_Index.hpp>
+
+#include <Kokkos_Core.hpp>
+
+#include <type_traits>
 
 namespace Cabana
 {
-namespace Experimental
-{
 //---------------------------------------------------------------------------//
 /*!
   \class RangePolicy
-  \brief Execution policy over a range of indices.
+  \brief Execution policy over a range of 1d indices.
+
+  Gives a linear range of indices for executing a functor with concurrency
+  over all indices.
 */
-template<int N, class ExecutionSpace>
-class RangePolicy
+template<class ExecutionSpace>
+class RangePolicy : public Kokkos::RangePolicy<ExecutionSpace>
 {
   public:
 
-    static constexpr int array_size = N;
+    using base_type = Kokkos::RangePolicy<ExecutionSpace>;
 
     using execution_space = ExecutionSpace;
 
-    // Range constructor.
+    /*!
+      \brief Range constructor.
+      \param begin The begininning of the 1D range.
+      \param begin The ending of the 1D range.
+    */
     RangePolicy( const std::size_t begin, const std::size_t end )
-        : _begin( begin )
-        , _end( end )
+        : base_type(begin,end)
+    {}
+
+    /*!
+      \brief Size constructor. Loop over n elements starting at 0.
+      \param n The number of elements in the range.
+    */
+    RangePolicy( const std::size_t n )
+        : base_type(0,n)
     {}
 
-    // Container constructor. The container must have a size() function that
-    // returns an int. C++ concepts would be really nice here. Valid
-    // containers include the AoSoA.
+    /*!
+      \brief Container constructor.
+
+      The container must have a size() function that returns an int. Valid
+      containers include the AoSoA, Slice, and Kokkos::View.
+
+      \param container The container over which to build the range policy.
+    */
     template<class Container>
     RangePolicy( Container container )
-        : _begin( 0 )
-        , _end( container.size() )
+        : base_type(0,container.size())
+    {}
+};
```

Comment:
I think you should get rid of this class altogether.  I don't see how neither the constructor from the extent of the range nor the one from a generic container bring enough added value to justify the existence of that class.

Reply:
For now I do like the idea of having two different objects to represent our 2 distinct parallel loop strategies over the AoSoA data structure. I know that one of them is exactly a `Kokkos::RangePolicy` (and I have renamed this to `Cabana::LinearPolicy`) but I like having a single basic class in which we can encapsulate our ideas for documentation purposes.

I suppose we could just bring `Kokkos::RangePolicy` into the Cabana namespace but then how would we document this as part of the API/Design of the library?

Reply:
By just calling it what it is and instead focus on spelling out what Cabana brings that's not already in Kokkos.

Reply:
OK I have removed range policy - people should use a `Kokkos::RangePolicy` with a `Kokkos::parallel_for` if they want to do linear loops over the data structure.